[Merged by Bors] - Add Experimental QUIC support

AgeManning opened the pull request:
## Issue Addressed

#4402 

## Proposed Changes

This PR adds QUIC support to Lighthouse. As this is not officially spec'd this will only work between lighthouse <-> lighthouse connections. We attempt a QUIC connection (if the node advertises it) and if it fails we fallback to TCP. 

This should be a backwards compatible modification. We want to test this functionality on live networks to observe any improvements in bandwidth/latency.

NOTE: This also removes the websockets transport as I believe no one is really using it. It should be mentioned in our release however.

AgeManning committed:
Initial quic support

jxs committed:
allow support for the two different UDP ports,

Discovery and Quic. Listen on both when starting th swarm.

jxs committed:
enable Quic transport on libp2p

AgeManning committed:
Merge pull request #5 from jxs/quic

or quic transport with tcp

AgeManning committed:
Add ENR structure for quick and specify listening criteria

jxs committed:
discovery: append quic multiaddrs along with tcp,

on handle_pending_outbound_connection

AgeManning committed:
Add some documentation for protocol developers

jxs committed:
discovery return Enr's instead of PeerId's

on GenerateEvent, and therefore remove unnecessary Enr fetching on Network.

jxs committed:
isolate multiaddresses gathering to dial on PeerManager

AgeManning committed:
Merge pull request #6 from jxs/quic

discovery: append quic multiaddrs along with tcp,

AgeManning committed:
Small logging and comment changes

AgeManning committed:
Merge latest unstable

AgeManning committed:
Correct line spacing in book docs

AgeManning committed:
Appease the all-wise all-knowing clippy

AgeManning committed:
fmt

AgeManning committed:
Correct quic port and improve logs

jxs committed:
Discovery:

- remove duplicated check for empty subnets
- remove no longer used `enr_for_peer`

jxs committed:
Network: cleanup remove no longer required fn

it's only a one liner for PeerManager delegation of discovered peers.

jxs committed:
PeerManager: move should_dial() logic to dial_peer()

as we seem to be calling everytime we call dial_peer().

jxs committed:
cargo fmt

AgeManning committed:
fmt

AgeManning committed:
Remove unused var

AgeManning committed:
Can't change a single line without fmt complaining

AgeManning committed:
Add QUIC to default ENR builder and fight tests

jxs committed:
Merge branch 'quic' of github.com:AgeManning/lighthouse into quic

jxs committed:
cargo fmt

jxs committed:
remove leftover println's

jxs committed:
fix benchmarks

mxinden commented on beacon_node/lighthouse_network/Cargo.toml:
```diff
@@ -44,11 +44,12 @@ prometheus-client = "0.21.0"
 unused_port = { path = "../../common/unused_port" }
 delay_map = "0.3.0"
 void = "1"
+libp2p-quic= { version = "0.9.0-alpha", features=["tokio"]}
```
```suggestion
libp2p-quic= { version = "0.9.1-alpha", features=["tokio"]}
```

Once https://github.com/libp2p/rust-libp2p/pull/4312 is merged and released.

mxinden replied:
Merged and released :heavy_check_mark:

jxs replied:
thanks Max, updated!

mxinden commented on beacon_node/lighthouse_network/src/service/utils.rs:
```diff
@@ -37,32 +39,42 @@ pub struct Context<'a> {
 
-    // Authentication
-    Ok((transport, bandwidth))
+    Ok((transport.boxed(), bandwidth))
```
`.with_bandwidth_logging` already calls `boxed`. Is the additional boxing needed here?

jxs replied:
nope, thanks Max addressed

mxinden reviewed (commented):
Excited for this to land.

Let me know in case there is anything I can do to help.

Ideally I would like to merge https://github.com/libp2p/rust-libp2p/pull/4304 before cutting a stable release. That said, happy to be convinced otherwise and cut a stable release of `libp2p-quic` earlier.

jxs committed:
fix debug tests

AgeManning committed:
Correct port mapping bug

AgeManning committed:
Merge latest unstable

jxs committed:
Merge branch 'quic' of github.com:AgeManning/lighthouse into quic

AgeManning committed:
Merge pull request #7 from jxs/quic

cleanups

jxs committed:
fix enr-quic6-port test

jxs committed:
update libp2p-quic to 0.9.2

jxs committed:
cargo clippy

jxs committed:
EnrExt: rename quic to quic4 for consistency

jxs committed:
increase quic port on dht persistance test

jxs committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into quic

jxs committed:
fix quic port on unused_v6_ports test

jxs committed:
make rpc tests use tcp multiaddr,

and rename them accordingly

jxs committed:
fix network_port_discovery_quic_port_flags_over_ipv4_and_ipv6 test

jxs committed:
fix test

network_port_discovery_quic_port_flags_over_ipv4_and_ipv6

jxs committed:
add quic to initialized ENR log

jxs committed:
add quic to ENR echo logs

jxs committed:
fix simulations by incrementing quic port by 1000

jxs committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into quic

jxs committed:
Merge branch 'quic' of github.com:AgeManning/lighthouse into quic

jxs committed:
add quic rpc tests.

jxs committed:
put quic transport in the dns transport

jxs committed:
add multiaddress to connection established logging

jxs committed:
update local test net scripts for quic

jxs committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into quic

jxs committed:
update libp2p-quic to stable

jxs committed:
add connected TCP and QUIC peers to prometheus metrics

AgeManning committed:
Differentiate the metrics in prometheus

jxs committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into quic

jxs committed:
Merge branch 'quic' of github.com:AgeManning/lighthouse into quic

jxs committed:
cargo fmt

jxs committed:
remove unrequired features section

jxs commented on beacon_node/lighthouse_network/src/discovery/mod.rs:
```diff
@@ -733,23 +722,6 @@ impl<TSpec: EthSpec> Discovery<TSpec> {
         target_peers: usize,
         additional_predicate: impl Fn(&Enr) -> bool + Send + 'static,
     ) {
-        // Make sure there are subnet queries included
```
`start_query` is only called with `QueryType::Subnet` on:
https://github.com/sigp/lighthouse/blob/96a381004ef9adf46bb3a01929be58485ea54192/beacon_node/lighthouse_network/src/discovery/mod.rs#L697-L712
which only calls `starty_query` if `!filtered_subnet_queries.is_empty()`

AgeManning replied:
Yes, I followed this logic when you proposed the changes. Looks good to me.

jxs commented on beacon_node/lighthouse_network/src/discovery/mod.rs:
```diff
@@ -957,23 +928,6 @@ impl<TSpec: EthSpec> NetworkBehaviour for Discovery<TSpec> {
     ) {
     }
 
-    fn handle_pending_outbound_connection(
```
multiaddress gathering is now isolated in PeerManager:
https://github.com/sigp/lighthouse/blob/cd66aeea317d6d1581f1136ca6a9b07ae750e9ea/beacon_node/lighthouse_network/src/peer_manager/network_behaviour.rs#L99-L113

see cd66aeea317d6d1581f1136ca6a9b07ae750e9ea

AgeManning replied:
Yep

jxs commented on beacon_node/lighthouse_network/src/peer_manager/mod.rs:
```diff
@@ -330,27 +326,25 @@ impl<TSpec: EthSpec> PeerManager<TSpec> {
-            if (min_ttl.is_some()
-                && connected_or_dialing + to_dial_peers.len() < self.max_priority_peers()
-                || connected_or_dialing + to_dial_peers.len() < self.max_peers())
-                && self.network_globals.peers.read().should_dial(&peer_id)
```
`should_dial` is done now inside `dial_peer` as it was being called every time `dial_peer` was also being called, see 191e0ae9c7a91ee5c89d4e774fcee542d8458789

AgeManning replied:
Yeah, nice

jxs commented on beacon_node/network/src/service.rs:
```diff
@@ -631,19 +627,6 @@ impl<T: BeaconChainTypes> NetworkService<T> {
                         warn!(self.log, "Failed to update ENR"; "error" => e);
                     }
                 }
-                // if the discovery service is not auto-updating, update it with the
```
removed on 61d4ae13af06366fb4123af081234812d2b5f3ab, is it not required?

AgeManning replied:
It was annoying to send back the UDP sockets with the added QUIC support. I realised the only reason we send back the UDP sockets was to update the ENR in this code section. After looking at it, i decided to remove it for two reasons. 

1. It was kinda annoying to send back all the udp-sockets and code was simpler this way. 
2. I don't think this behaviour is intuitive and actually what we want. The `auto-update` feature of discovery is on by default. Users should only ever remove that if that know what they are doing and they know their own external ports and that they are not going to change. So if a user has selected that option and set specific external ports, I dont think we should change those settings because UPnP may come around and give us new ports. i.e if you set auto-update to be off, your ports are static. Removing this code makes them static, which I think is how it should be.

jxs replied:
thanks for explaining Age, makes sense!

divagant-martian replied:
Agree it makes sense to remove the updating of dicovery sockets, but you would still need to do this for QUIC sockets since discovery updates are unrelated to them

jxs committed:
move parse_listening_addresses for better reviewing

jxs committed:
improve doc wording

jxs reviewed (approved):
only one question left!

jxs committed:
fix metrics protocol identification

divagant-martian commented on beacon_node/lighthouse_network/src/discovery/enr.rs:
```diff
@@ -163,7 +165,27 @@ pub fn create_enr_builder_from_config<T: EnrKey>(
-    if enable_tcp {
+    if enable_libp2p {
+        // Add QUIC fields to the ENR.
+        // If `enable_libp2p` is disabled, then we should not support QUIC in the ENR either.
```
While I understand the logic behind this, the comment is not very explanatory. 

Maybe something like

```
Since QUIC is used as an alternative transport for the libp2p protocols, the related fields should only be added when both QUIC and libp2p are enabled
```

jxs replied:
agreed, updated Thanks!

divagant-martian commented on beacon_node/lighthouse_network/src/discovery/enr_ext.rs:
```diff
@@ -174,6 +209,31 @@ impl EnrExt for Enr {
     }
 
+    /// Returns a list of multiaddrs if the ENR has an `ip` and a `quic` key **or** an `ip6` and a `quic6`.
+    /// The vector remains empty if these fields are not defined.
```
Seems unnecessary?

divagant-martian commented on beacon_node/lighthouse_network/src/discovery/enr_ext.rs:
```diff
@@ -174,6 +209,31 @@ impl EnrExt for Enr {
+    /// The vector remains empty if these fields are not defined.
+    fn multiaddr_quic(&self) -> Vec<Multiaddr> {
+        let mut multiaddrs: Vec<Multiaddr> = Vec::new();
+        // Check for quic first as it is less likely.
```
This comment seems out of place for a `multiaddr_quic` fn

jxs replied:
agreed, removed it.

divagant-martian commented on beacon_node/lighthouse_network/src/listen_addr.rs:
```diff
@@ -46,22 +55,41 @@ impl ListenAddress {
-    pub fn tcp_addresses(&self) -> impl Iterator<Item = Multiaddr> + '_ {
-        let v4_multiaddr = self
+    /// Returns the addresses the Swarm will listen on, given the setup.
+    pub fn listen_addresses(&self) -> impl Iterator<Item = Multiaddr> {
```
we still listen on the discovery addresses, maybe `libp2p_addresses` would be more correct?

divagant-martian commented on beacon_node/lighthouse_network/src/peer_manager/network_behaviour.rs:
```diff
@@ -234,25 +253,48 @@ impl<TSpec: EthSpec> PeerManager<TSpec> {
-        }
+        };
 
         // increment prometheus metrics
+        match remote_addr.iter().find(|proto| {
```
missing check for metrics enabled

jxs replied:
thanks, updated

divagant-martian commented on beacon_node/lighthouse_network/src/peer_manager/network_behaviour.rs:
```diff
@@ -278,7 +320,27 @@ impl<TSpec: EthSpec> PeerManager<TSpec> {
+            ConnectedPoint::Dialer { address, .. } => address,
+        };
+
         // Update the prometheus metrics
+        match remote_addr.iter().find(|proto| {
```
missing check for metrics enabled

jxs replied:
thanks, updated

divagant-martian commented on beacon_node/lighthouse_network/src/service/mod.rs:
```diff
@@ -454,6 +462,21 @@ impl<AppReqId: ReqId, TSpec: EthSpec> Network<AppReqId, TSpec> {
+                }
+            }
+
+            //
```
```suggestion
```

jxs replied:
thanks, updated

divagant-martian commented on beacon_node/lighthouse_network/src/service/mod.rs:
```diff
@@ -1036,31 +1059,27 @@ impl<AppReqId: ReqId, TSpec: EthSpec> Network<AppReqId, TSpec> {
+        // Remove the ENR from the cache to prevent continual re-dialing on disconnects
+        peers_to_dial.iter().for_each(|enr| {
+            self.peer_manager_mut().dial_peer(enr.clone());
+            self.discovery_mut().remove_cached_enr(&enr.peer_id());
```
any reason behind doing this in this order? it looks to me like this could be done consuming the  iterator (`into_iter`), calling discovery first and then passing the enr saving the clones

On a related note, as far as I remember a for loop is more efficient than a `for_each` call

Also, why remove the log? too noisy?

jxs replied:
thanks for the suggestion Diva, 
`cached_enrs` returns `&Enr` so you have to clone at least when calling `dial_peer` but by maintaining the previous order as you suggested we are able to save a clone. Regarding for loop, agree that it's more idiomatic updated it, but performance wise vs iterators it can [actually be the opposite](https://doc.rust-lang.org/book/ch13-04-performance.html). 

The log was mistakenly removed on cd66aeea317d6d1581f1136ca6a9b07ae750e9ea I re-added it thanks

divagant-martian commented on beacon_node/lighthouse_network/tests/common.rs:
```diff
@@ -132,14 +138,28 @@ pub async fn build_node_pair(
+            .multiaddr_quic()
+            .first()
+            .unwrap()
+            .clone(),
```
barely a nit, but being these tests, why not just `pop` which returns by value instead of by reference?

```suggestion
            .pop()
            .unwrap(),
        Protocol::Quic => receiver
            .local_enr()
            .multiaddr_quic()
            .pop()
            .unwrap(),
```

jxs replied:
makes sense, thanks. Updated

divagant-martian commented on beacon_node/lighthouse_network/tests/rpc_tests.rs:
```diff
@@ -836,10 +880,80 @@ fn test_blocks_by_root_chunked_rpc_terminates_correctly() {
+                panic!("Future timed out");
+            }
+        }
+    })
```
all this code is repeated. Why not move it to a fn that received the transport to use instead?

divagant-martian commented on beacon_node/src/cli.rs:
```diff
@@ -225,10 +257,19 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
             Arg::with_name("disable-discovery")
                 .long("disable-discovery")
                 .help("Disables the discv5 discovery protocol. The node will not search for new peers or participate in the discovery protocol.")
+                .takes_value(false)
```
is this necessary?

jxs replied:
don't think so, I think `takes_value` by default is false, updated it thanks. There are some others left previously, can be addressed on another PR

divagant-martian reviewed (commented):
looks generally straight forward. Have you people checked the RAM increase with this change? how does it look like?

jxs committed:
update socket address for quic connections,

on update_connection_state()

jxs committed:
address review

AgeManning committed:
Remove references and assumptions to TCP based libp2p port

AgeManning committed:
fmt

jxs committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into quic

AgeManning committed:
Clippy lints

jxs committed:
Merge branch 'quic' of github.com:AgeManning/lighthouse into quic

AgeManning committed:
Improve inbound connection log

AgeManning committed:
Update QUIC ports on UPnP mapping changes

AgeManning committed:
Merge latest unstable

jxs committed:
fix quic metric protocol on connection close

jxs committed:
Merge branch 'quic' of github.com:AgeManning/lighthouse into quic

jimmygchen commented on scripts/local_testnet/start_local_testnet.sh:
```diff
@@ -139,7 +139,7 @@ sed -i 's/"shanghaiTime".*$/"shanghaiTime": 0,/g' $genesis_file
     secret=$DATADIR/geth_datadir$bn/geth/jwtsecret
     echo $secret
-    execute_command_add_PID beacon_node_$bn.log ./beacon_node.sh $SAS -d $DEBUG_LEVEL $DATADIR/node_$bn $((BN_udp_tcp_base + $bn)) $((BN_http_port_base + $bn)) http://localhost:$((EL_base_auth_http + $bn)) $secret
+    execute_command_add_PID beacon_node_$bn.log ./beacon_node.sh $SAS -d $DEBUG_LEVEL $DATADIR/node_$bn $((BN_udp_tcp_base + $bn)) $((BN_http_port_base + $bn)) $((BN_udp_tcp_base + $bn + 100)) http://localhost:$((EL_base_auth_http + $bn)) $secret
```
This change also needs to be made in the `doppelganger_protection` script as it uses the same `beacon_node.sh` script.

https://github.com/sigp/lighthouse/blob/70c4ae35aba72e86884143ba054012e104008c04/scripts/tests/doppelganger_protection.sh#L58-L60

jxs committed:
fix doppelganger_protection script

jxs committed:
fix ports for doppelganger protection test on ci

jxs committed:
fix missing port for last validator client

jimmygchen commented on scripts/local_testnet/start_local_testnet.sh:
```diff
@@ -139,7 +139,7 @@ sed -i 's/"shanghaiTime".*$/"shanghaiTime": 0,/g' $genesis_file
     secret=$DATADIR/geth_datadir$bn/geth/jwtsecret
     echo $secret
-    execute_command_add_PID beacon_node_$bn.log ./beacon_node.sh $SAS -d $DEBUG_LEVEL $DATADIR/node_$bn $((BN_udp_tcp_base + $bn)) $((BN_http_port_base + $bn)) http://localhost:$((EL_base_auth_http + $bn)) $secret
+    execute_command_add_PID beacon_node_$bn.log ./beacon_node.sh $SAS -d $DEBUG_LEVEL $DATADIR/node_$bn $((BN_udp_tcp_base + $bn)) $((BN_http_port_base + $bn)) $((BN_udp_tcp_base + $bn + 100)) http://localhost:$((EL_base_auth_http + $bn)) $secret
```
looks like the args order is different to what's expected in the `beacon_node.sh` script, it's expecting ports in this order: `tcp_port`, `quic_port`, `http_port`, so it should probably be:

```suggestion
    execute_command_add_PID beacon_node_$bn.log ./beacon_node.sh $SAS -d $DEBUG_LEVEL $DATADIR/node_$bn $((BN_udp_tcp_base + $bn)) $((BN_udp_tcp_base + $bn + 100)) $((BN_http_port_base + $bn)) http://localhost:$((EL_base_auth_http + $bn)) $secret
```

jxs replied:
yup right, thanks Jimmy!

jimmygchen commented on scripts/tests/doppelganger_protection.sh:
```diff
@@ -96,7 +96,7 @@ if [[ "$BEHAVIOR" == "success" ]]; then
     echo "Starting the last validator client"
 
-    ../local_testnet/validator_client.sh $HOME/.lighthouse/local-testnet/node_4 http://localhost:8100 &
+    ../local_testnet/validator_client.sh $HOME/.lighthouse/local-testnet/node_4 http://localhost:9100 &
```
ah looks like there are two more beacon API calls here:

https://github.com/sigp/lighthouse/blob/dd444767c66645f77a10132f97e0d73ae0d77f88/scripts/tests/doppelganger_protection.sh#L113

and 

https://github.com/sigp/lighthouse/blob/dd444767c66645f77a10132f97e0d73ae0d77f88/scripts/tests/doppelganger_protection.sh#L131

jimmygchen replied:
^ needs to be updated to 9100 too

jxs replied:
thanks Jimmy, addressed.

jxs committed:
fix last two ports in doppelganger protection test

jxs committed:
fix port order on start_local_testnet.sh

AgeManning committed:
Hopefully improve port management in rpc tests

AgeManning committed:
fmt

AgeManning commented:
bors r+

AgeManning committed:
Merge latest unstable

AgeManning committed:
Update tests due to unstable merge

divagant-martian commented on beacon_node/lighthouse_network/src/listen_addr.rs:
```diff
@@ -6,17 +6,26 @@ use serde::{Deserialize, Serialize};
 /// A listening address composed by an Ip, an UDP port and a TCP port.
```
worth adding quic here in the docs

divagant-martian commented on beacon_node/lighthouse_network/src/listen_addr.rs:
```diff
@@ -6,17 +6,26 @@ use serde::{Deserialize, Serialize};
+        (self.addr.clone().into(), self.quic_port).into()
+    }
+
+    pub fn libp2p_socket_addr(&self) -> SocketAddr {
```
```suggestion
    pub fn tcp_socket_addr(&self) -> SocketAddr {
```
so I realized I suggested `libp2p` over `listening` but this change doesn't seem correct to me either. 

`quic` and `tcp` are both transports over which `libp2p` can work. A `libp2p` socket address could be a `quic` one, whereas `tcp` and `quic` unequivocally refer to what we are trying to convey

divagant-martian commented on beacon_node/lighthouse_network/src/listen_addr.rs:
```diff
@@ -46,22 +55,41 @@ impl ListenAddress {
-    pub fn tcp_addresses(&self) -> impl Iterator<Item = Multiaddr> + '_ {
-        let v4_multiaddr = self
+    /// Returns the addresses the Swarm will listen on, given the setup.
+    pub fn libp2p_addresses(&self) -> impl Iterator<Item = Multiaddr> {
```
this change looks right to me and follows the same logic of why the above would be better kept as `tcp`. Otherwise, these two have different meanings

divagant-martian commented on beacon_node/lighthouse_network/tests/rpc_tests.rs:
```diff
@@ -833,14 +877,9 @@ fn test_blocks_by_root_chunked_rpc_terminates_correctly() {
-    let log_level = Level::Trace;
-    let enable_logging = false;
-
+/// Established a pair of nodes and disconnects the pair based on the selected protocol via an RPC
```
```suggestion
/// Establishes a pair of nodes and disconnects the pair based on the selected protocol via an RPC
```

divagant-martian commented:
Minor nits, sorry if I'm late to the party but curiosity won

An exciting change you two! :tada:

divagant-martian commented on beacon_node/lighthouse_network/tests/common.rs:
```diff
@@ -67,15 +68,34 @@ pub fn build_log(level: slog::Level, enabled: bool) -> slog::Logger {
     let mut config = NetworkConfig::default();
+
+    // Find unused ports
+    let tcp_port = unused_tcp4_port().unwrap();
+    let disc_port = unused_udp4_port().unwrap();
+    let quic_port = unused_udp4_port().unwrap();
```
just realized that this undoes #4705 partially, since it does not use zero ports for the rpc tests anymore

divagant-martian commented on beacon_node/lighthouse_network/tests/common.rs:
```diff
@@ -67,15 +68,34 @@ pub fn build_log(level: slog::Level, enabled: bool) -> slog::Logger {
+        disc_port,
+        quic_port,
+    );
+    config.enr_udp4_port = if disc_port == 0 {
+        None
+    } else {
+        Some(disc_port)
+    };
```
the `None` here is most likely unreachable

AgeManning committed:
Small test changes

AgeManning commented:
bors r+

AgeManning commented:
bors retry

AgeManning commented:
bors retry

AgeManning commented:
bors retry

AgeManning commented:
bors retry

AgeManning committed:
Nuke windows self-hosted runner

AgeManning commented:
bors r+

AgeManning commented:
bors retry

AgeManning commented:
bors r+

AgeManning committed:
Revert "Nuke windows self-hosted runner"

This reverts commit 231e7dbf6e26eb113713b4a371224c107eec6035.

divagant-martian commented:
![image](https://github.com/sigp/lighthouse/assets/26765164/2aac0cde-a355-4665-a5a1-39f19e3d59fc)

AgeManning commented:
bors r+

AgeManning commented:
bors cancel :(

jxs committed:
beacon node: move connected peer metrics,

put them before the ban status checks to avoid negative counts as each FromSwarm::ConnectionClosed
event is always paired with an earlier FromSwarm::ConnectionEstablished, and when we ban before
incrementing metrics we return earlier.

AgeManning commented:
bors r+

AgeManning commented:
bors r+
